### PR TITLE
.detach_() pytorch tensor to serialize data as numpy array.

### PR DESCRIPTION
**Pull request by muammar**
This commit is related to #2581. Tensors with requires_grad=True cannot
be converted to numpy unless they are detached from the computational
graph first. To fix that, the tensor has to be detached as a leaf using
the .detach_() method. For more information see:

https://pytorch.org/docs/stable/autograd.html#torch.Tensor.detach_

### CODE REVIEW

**Comment by mrocklin** [comment]
Thanks for the PR @muammar !

Can I ask you to include your example in the issue as a small test?  See https://github.com/dask/distributed/blob/master/distributed/protocol/tests/test_torch.py for examples.

Also cc @stsievert

**Comment by muammar** [comment]
I hope I did this correctly. Running `py.test` called the function correctly.

**Comment by stsievert** [comment]
I'd like to see `requires_grad` tested. Something like

``` python
x = torch.tensor([0, 0, 0], requires_grad=True)
y = deserialize(*serialize(x))
assert y.requires_grad
```

What's the motivation behind this PR? Are models being trained with PyTorch and Dask?

File: distributed/protocol/tests/test_torch.py [hunk 1]
```diff
@@ -31,3 +31,10 @@ def test_resnet():
     header, frames = serialize(model)
     model2 = deserialize(header, frames)
     assert str(model) == str(model2)
+
+
+def test_deserialize_grad():
+    torch.set_printoptions(precision=10)
+    a = np.random.rand(8, 1)
+    t = torch.tensor(a, requires_grad=True, dtype=torch.float)
+    deserialize(*serialize(t))
```
Comment (mrocklin) [inline distributed/protocol/tests/test_torch.py]: The result of this line should be a PyTorch tensor that is mostly equivalent to `t`.  Is there some check we can run to verify that we have recreated an equivalent tensor object?

File: distributed/protocol/tests/test_torch.py [hunk 2]
```diff
@@ -31,3 +31,10 @@ def test_resnet():
     header, frames = serialize(model)
     model2 = deserialize(header, frames)
     assert str(model) == str(model2)
+
+
+def test_deserialize_grad():
+    torch.set_printoptions(precision=10)
```
Comment (mrocklin) [inline distributed/protocol/tests/test_torch.py]: My guess is that this line was used for debugging.  Maybe we can remove it before we merge?

File: distributed/protocol/tests/test_torch.py (hunk 1 above)
Comment (muammar) [reply distributed/protocol/tests/test_torch.py]: By doing this: 

```
import torch
import numpy as np
from distributed.protocol import serialize, deserialize

a = np.random.rand(8, 1)

t = torch.tensor(a, requires_grad=True, dtype=torch.float)

t2 = deserialize(*serialize(t))
print(a)
print(t2.numpy())

```

I get: 
```
[[0.90611823]
 [0.57207592]
 [0.14868348]
 [0.24763735]
 [0.25475143]
 [0.71949837]
 [0.19011521]
 [0.3211044 ]]
[[0.9061182 ]
 [0.5720759 ]
 [0.14868347]
 [0.24763735]
 [0.25475144]
 [0.7194984 ]
 [0.19011521]
 [0.3211044 ]]
```

They slightly differ on the 8th decimal. `assert` will thorw an error because of that and I thought that setting larger precission would help.

Comment (muammar) [reply distributed/protocol/tests/test_torch.py]: Getting to check if they are equivalent, I could use https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.around.html#numpy.around. And truncate to the 7th or 6th decimal.

**Comment by muammar** [comment]
@stsievert that is correct. I am trying to train a model using pytorch and dask.

**Comment by muammar** [comment]
> I'd like to see `requires_grad` tested. Something like
> 
> ```python
> x = torch.tensor([0, 0, 0], requires_grad=True)
> y = deserialize(*serialize(x))
> assert y.requires_grad
> ```

Thanks for bringing this up. The data is being serialized/deserialized correctly. However, `.detach_()` creates a copy of the tensor with `requires_grad=False`. Shouldn't we use `.detach()` instead? because how else we could verify the tensor requires grad.

**Comment by stsievert** [comment]
> `.detach_()` creates a copy of the tensor with `requires_grad=False`. Shouldn't we use `.detach()` instead? because how else we could verify the tensor requires grad.

The `requires_grad` information will need be encoded, probably in the header. Maybe

``` python
def serialize_torch(x):
    # ...
    if x.requires_grad:
	    kwargs["requires_grad"] = True
        x.requires_grad_()
    header, frames = serialize(x.numpy())
    header.update(kwargs)
    # ...
```

with a matching thing for `deserialize` that looks at the `header`.

> I am trying to train a model using pytorch and dask.

I've had a good experience training using PyTorch models with Dask when using [skorch] (the official wrapper [ignite] might work too, but I'm not sure).

[skorch]:https://skorch.readthedocs.io/en/latest/
[ignite]:https://pytorch.org/ignite/

**Comment by muammar** [comment]
> > `.detach_()` creates a copy of the tensor with `requires_grad=False`. Shouldn't we use `.detach()` instead? because how else we could verify the tensor requires grad.
> 
> The `requires_grad` information will need be encoded, probably in the header. Maybe
> 
> ```python
> def serialize_torch(x):
>     # ...
>     if x.requires_grad:
> 	    kwargs["requires_grad"] = True
>         x.requires_grad_()
>     header, frames = serialize(x.numpy())
>     header.update(kwargs)
>     # ...
> ```
> 
> with a matching thing for `deserialize` that looks at the `header`.
> 

Inspired by your comment I came up with https://github.com/dask/distributed/pull/2586/commits/d9143a0e200a6ed5ea67be51bc8844f1b3a6357d. It works correctly for `requires_grad` true and false. 

> > I am trying to train a model using pytorch and dask.
> 
> I've had a good experience training using PyTorch models with Dask when using [skorch](https://skorch.readthedocs.io/en/latest/) (the official wrapper [ignite](https://pytorch.org/ignite/) might work too, but I'm not sure).

Thanks for sharing!. I will look into it, but would be very glad if I would not need to add another dependency.

File: distributed/protocol/tests/test_torch.py (hunk 1 above)
Comment (stsievert) [reply distributed/protocol/tests/test_torch.py]: > They slightly differ on the 8th decimal

I use `np.allclose` in these situations.

Comment (muammar) [reply distributed/protocol/tests/test_torch.py]: @stsievert that worked. Thanks for the suggestion.

**Comment by mrocklin** [comment]
@stsievert does this look ok to you?  If so, I'm going to merge this in.

**Comment by stsievert** [comment]
> does this look ok to you?

Yes. This PR more accurately represents the tensor that's communicated. The `require_grad` flag specifies that the gradients should be recorded when performing `loss.backwards()`. 

The gradients do eventually need to be connected to an optimizer. @muammar implemented a parameter server for that (https://github.com/dask/distributed/issues/2581#issuecomment-479024638).

**Comment by mrocklin** [comment]
Thanks @muammar .  This is in!

Also, I notice that this is your first code contribution to this repository.  Welcome!

**Comment by muammar** [comment]
> Thanks @muammar . This is in!
> 
> Also, I notice that this is your first code contribution to this repository. Welcome!


You are right, this is my first contribution to dask!! thanks for accepting the PR.
